### DISCUSSION TITLE
`sa.orm.relationship(uselist=False)` should have `LIMIT 1` in sql

### DISCUSSION THREAD

**Discussion by loynoir** [sqlalchemy/sqlalchemy#7638] [Usage Questions] [answered]
### Describe the bug

`sa.orm.relationship(uselist=False)` should have `LIMIT 1` in sql

### To Reproduce

Turn on sqlalchemy logging, and

```python
    field = sa.orm.relationship(
        xxx,
        primaryjoin=  sa.and_(
            xxx,
            xxx
        ),
        order_by = sa.desc(xxx),
        uselist=False
    )
```

### Actual sql
```sql
BEGIN;

SELECT
    {{every_fields}}
FROM
    {{table_name}}
WHERE
    {{table_name}}.{{table_id}} = ?;

SELECT
    {{every_fields}}
FROM
    {{table_name}}
WHERE
    {{primaryjoin}}
ORDER BY
    {{table_name}}.{{order_by}} DESC;
```

### Expected sql
```sql
BEGIN;

SELECT
    {{every_fields}}
FROM
    {{table_name}}
WHERE
    {{table_name}}.{{table_id}} = ?;

SELECT
    {{every_fields}}
FROM
    {{table_name}}
WHERE
    {{primaryjoin}}
ORDER BY
    {{table_name}}.{{order_by}} DESC
LIMIT 1;
```

### Versions

- OS:
- Python:
- SQLAlchemy: 1.4.31
- Database: sqlite3
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): sqlite3

### Additional context
1. May it be possible to customize `select {{every_fields}}`?

2. More generically, how can access ORM's SQL expression? So, all `select(partial_fields)`, `limit(n)`, `order_by(expr)` can be combined?

3. More generically, how can access ORM's output?

DB -> SQL -> sqlalchemy ORM -> raw ORM object -> **custom hook?** -> final property

**Comment by zzzeek** [comment, marked as answer]
using the uselist=False flag with a relationship that refers to a remote table where the foreign key constraint is on the remote table is in SQLAlchemy known as the [one to one](https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#one-to-one).   The default version of this pattern assumes that the database itself maintains exactly one row for the related object, and as you might have noticed, if you fetch this relationship and more than one row is returned, there's a warning emitted.  this warning suggests that either the state of the database is not what's expected, or that the relationship is not configured in the correct way such that one row is returned.

that is, if you want a relationship to a SELECT with LIMIT, that's a thing you'd need to explicitly implement.   it's in fact not that simple to apply LIMIT on the relationship in a generic way.     For example.  Suppose you were using [joined eager loading](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html?highlight=joinedload#sqlalchemy.orm.joinedload).  in this form of loading, to load a list of A rows with related B rows looks like:

     SELECT a.*, b.* FROM a LEFT OUTER JOIN b ON a.id == b.a_id

can we just stick LIMIT 1 there?  not really:

     SELECT a.*, b.* FROM a LEFT OUTER JOIN b ON a.id == b.a_id LIMIT 1

above we see the LIMIT would be ruining the query.   The use case you have above is specific to just one type of loading called [lazy loading](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html?highlight=joinedload#lazy-loading), but relationship() can't map something that only works with one kind of loading.

So the normal way to map to a relationship where you want the related items to be on a LIMIT is to map the relationship to a subquery of that item.     when we do that in a simplistic way, we end up with a subquery that has to use correlation to the parent table, which is fine but not the most efficient approach (I dont have an example handy at the moment).  but long story short we have an approach that does this in the most efficient way possible using SQL window functions, which you can see at https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#row-limited-relationships-with-window-functions .   Most major databases support window functions so I would suggest dropping in this pattern for those cases where you want a relationship to link to a set of items where you want only the first one.  I'll adapt it to uselist=False here:

```py
from sqlalchemy import and_
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import func
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import aliased
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session

Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)

    # list of B, just for persistence
    bs = relationship("B")


class B(Base):
    __tablename__ = "b"
    id = Column(Integer, primary_key=True)
    a_id = Column(ForeignKey("a.id"))


partition = select(
    B,
    func.row_number().over(order_by=B.id, partition_by=B.a_id).label("index"),
).alias()

partitioned_b = aliased(B, partition)

# scalar B relationship
A.b_scalar = relationship(
    partitioned_b,
    primaryjoin=and_(partitioned_b.a_id == A.id, partition.c.index < 2),
    uselist=False,
)


# demo

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

b1, b2, b3 = B(), B(), B()
a1 = A(bs=[b1, b2, b3])
s.add(a1)
s.commit()

assert a1.b_scalar is b1

```

**Reply by loynoir** [reply]
Oh, thank you very much for explaination and maintain `sqlalchemy`!!!

I bet I can never hand write that complicated sql!!! 👍

**Reply by loynoir** [reply]
@zzzeek 
1. Is it possible to not query all fields in `relationship()`, as I saw querying all fields in logging?

2. And let 

```python
obj.field = custom_hook(raw_orm_object)
```

instead of
```python
obj.field = raw_orm_object
```

**Reply by CaselIT** [reply]
regarding 1 there is `deferred`. it's explained in the documentation. If you want to instead control it in a per query you can use the load_only relationship loader option

not sure what you are asking in the second question

**Reply by loynoir** [reply]
@CaselIT 

Given
```python
assert a1.b_scalar is b1
```

Is it possible having `relationship_hook_on_A_bs`
```python
def relationship_hook_on_A_bs(b):
    if b is b1:
        return 42
    else:
        return 0


assert a1.b_scalar == 42
```

**Reply by CaselIT** [reply]
I don't think so. If you want to access an attribute on b from a you could use `association_prpxy`

Otherwise you could use a normal `@property` on the A class
